contracts/protocol/tokenization/NTokenLiquidity.sol

```diff
@@ -0,0 +1,204 @@
+     */
+    constructor(IPool pool, address delegateRegistry)
+        NToken(pool, true, delegateRegistry)
+    {
+        _ERC721Data.balanceLimit = 30;
+    }
+
+    function _underlyingAsset(address, uint256)
+        internal
+        view
+        virtual
+        returns (address, address)
+    {
+        return (address(0), address(0));
+    }
+
+    function _decreaseLiquidity(
+        address,
+        uint256,
+        uint128,
+        uint256,
+        uint256
+    ) internal virtual {}
+
+    function _increaseLiquidity(
+        address,
+        uint256,
+        uint256,
+        uint256,
+        uint256,
+        uint256
+    ) internal virtual returns (uint256, uint256) {
+        return (0, 0);
+    }
+
+    function _collect(address, uint256)
+        internal
+        virtual
+        returns (uint256, uint256)
+    {
+        return (0, 0);
+    }
+
+    function _refundETH(address) internal virtual {}
+
+    function underlyingAsset(uint256 tokenId)
+        external
+        view
+        returns (address token0, address token1)
+    {
+        return _underlyingAsset(_ERC721Data.underlyingAsset, tokenId);
+    }
+
+    /// @inheritdoc INTokenLiquidity
+    function decreaseLiquidity(
+        address user,
+        uint256 tokenId,
+        uint128 liquidityDecrease,
+        uint256 amount0Min,
+        uint256 amount1Min
+    )
+        external
+        onlyPool
+        nonReentrant
+        returns (
+            address token0,
+            address token1,
+            uint256 amount0,
+            uint256 amount1
+        )
+    {
+        require(user == ownerOf(tokenId), Errors.NOT_THE_OWNER);
+
+        address positionManager = _ERC721Data.underlyingAsset;
+        if (liquidityDecrease > 0) {
+            _decreaseLiquidity(
+                positionManager,
+                tokenId,
+                liquidityDecrease,
+                amount0Min,
+                amount1Min
+            );
+        }
+
+        (token0, token1) = _underlyingAsset(positionManager, tokenId);
+
+        (amount0, amount1) = _collect(positionManager, tokenId);
+    }
+
+    /// @inheritdoc INTokenLiquidity
+    function increaseLiquidity(
+        address payer,
+        uint256 tokenId,
+        uint256 amountAdd0,
+        uint256 amountAdd1,
+        uint256 amount0Min,
+        uint256 amount1Min
+    ) external payable onlyPool nonReentrant {
+        address positionManager = _ERC721Data.underlyingAsset;
+
+        (address token0, address token1) = _underlyingAsset(
+            positionManager,
+            tokenId
+        );
+
+        // move underlying into this contract
+        bool token0IsETH;
+        bool token1IsETH;
+        if (msg.value > 0) {
+            address weth = _addressesProvider.getWETH();
+            require(token0 == weth || token1 == weth, Errors.INVALID_AMOUNT);
+            token0IsETH = (token0 == weth);
+            token1IsETH = (token1 == weth);
+        }
```

Comment:
there are many duplicated code here, I think we can declare variable to represent ethToken, ethAmount